spark/src/main/scala/org/apache/druid/spark/clients/DruidClient.scala

```diff
@@ -0,0 +1,194 @@
+    JodaUtils.MIN_INSTANT,
+    JodaUtils.MAX_INSTANT
+  ))
+
+  /**
+    * The SQL system catalog tables are incorrect for multivalue columns and don't have accurate
+    * type info, so we need to fall back to segmentMetadataQueries. Given a DATASOURCE and a range
+    * of INTERVALS to query over, return a map from column name to a tuple of
+    * (columnType, hasMultipleValues). Note that this is a very expensive operation over large
+    * numbers of segments. If possible, this method should only be called with the most granular
+    * starting and ending intervals instead of over a larger interval.
+    *
+    * @param dataSource The Druid dataSource to fetch the schema for.
+    * @param intervals  The intervals to return the schema for, or None to query all segments.
+    * @return A map from column name to data type and whether or not the column is multi-value
+    *         for the schema of DATASOURCE.
+    */
+  def getSchema(dataSource: String, intervals: Option[List[Interval]]): Map[String, (String, Boolean)] = {
+    val queryInterval = intervals.getOrElse(DefaultSegmentMetadataInterval)
+    val body = Druids.newSegmentMetadataQueryBuilder()
+      .dataSource(dataSource)
+      .intervals(queryInterval.asJava)
+      .analysisTypes(SegmentMetadataQuery.AnalysisType.SIZE)
+      .merge(true)
+      .context(Map[String, AnyRef](
+        "timeout" -> Int.box(timeoutWaitMilliseconds)
+      ).asJava)
+      .build()
+    val response = sendRequestWithRetry(
+      druidBaseQueryURL(hostAndPort), numRetries, Option(MAPPER.writeValueAsBytes(body))
+    )
+    val segments =
+      MAPPER.readValue[JList[SegmentAnalysis]](
+        response.getContent, new TypeReference[JList[SegmentAnalysis]]() {})
+    if (segments.size() == 0) {
+      throw new IAE(
+        s"No segments found for intervals [${intervals.mkString(",")}] on $dataSource"
+      )
+    }
+    // Since we're setting merge to true, there should only be one item in the list
+    if (segments.size() > 1) {
+      throw new ISE("Merged segment metadata response had more than one row!")
+    }
+    log.debug(segments.asScala.map(_.toString).mkString("SegmentAnalysis: [", ", ", "]"))
+    /*
+     * If a dimension has multiple types within the spanned interval, the resulting column
+     * analysis will have the type "STRING" and be an error message. We abuse that here to infer
+     * a string type for the dimension and widen the type for the resulting DataFrame.
+     */
+    val columns = segments.asScala.head.getColumns.asScala.toMap
+    columns.foreach{ case(key, column) =>
+      if (column.isError) {
+        log.warn(s"Multiple column types found for dimension $key in interval" +
+          s" ${queryInterval.mkString("[", ", ", "]")}! Falling back to STRING type")
+      }
+    }
+    columns.map{ case (name: String, col: ColumnAnalysis) =>
+      name -> (col.getType, col.isHasMultipleValues)
+    }
+  }
+
+  private def sendRequestWithRetry(

```

Comment:
I had intentionally left the method unannotated because tail recursion doesn't do much here and it didn't seem worth confusing future developers who may be less familiar with Scala. Since it makes more sense to document why instead of not doing, I've added the annotation and a comment explaining why other people shouldn't worry if they need to change the behavior in the future.

Reply:
Cool, sounds good to me